.github/workflows/test.yml

```diff
@@ -0,0 +1,96 @@
+name: Test Build
+
+on:
+  push:
+    branches:
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+
+jobs:
+  test:
+    name: Gatsby + Craft with PHP ${{ matrix.php-versions }} + ${{ matrix.db }}
+    runs-on: ${{ matrix.operating-system }}
+    strategy:
+      matrix:
+        operating-system: [ubuntu-latest]
+        php-versions: ["7.4"]
+        db: ["mysql", "pgsql"]
+    services:
+      mysql:
+        image: mysql:5.7
+        env:
+          MYSQL_ALLOW_EMPTY_PASSWORD: false
+          MYSQL_ROOT_PASSWORD: root
+          MYSQL_DATABASE: test
+        ports:
+          - 3306:3306
+        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
+      postgres:
+        image: postgres:12
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: test
+        ports:
+          - 5432:5432
+        options: --health-cmd="pg_isready" --health-interval=10s --health-timeout=5s --health-retries=3
+      nginx:
+        image: craftcms/nginx:7.4
+        options: --volume ${{ github.workspace }}:/app --user www-data:www-data
+        ports:
+          - 80:8080
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Setup PHP Action
+        uses: shivammathur/setup-php@2.5.0
+        with:
+          php-version: ${{ matrix.php-versions }}
+          extensions: mbstring, intl, gd, imagick, zip, dom, pdo_mysql, pdo_pgsql, fileinfo
+          ini-values: post_max_size=256M
+          tools: php-cs-fixer, phpunit
+
+      - name: Prepare Craft environment
+        run: cp .github/workflows/.env.${{ matrix.db }} .env
+
+      - name: Validate composer.json and composer.lock
+        run: composer validate
+
+      - name: Install Composer dependencies
+        run: composer install --prefer-dist --no-progress
+      
+      - name: Add Docker database host aliases to build environment
+        run: sudo echo "127.0.0.1 nginx postgres mysql" | sudo tee -a /etc/hosts
+
+      - name: Install Craft
+        run: php craft install/craft --username=admin --password=password --email=test@craftcms.com --site-name=Craft Blog --site-url=$DEFAULT_SITE_URL --interactive=0
+
+      - name: Add Craft demo content
+        run: php craft migrate/up --track=content --interactive=0
+
+      - name: Set Craft writable folder permissions
+        run: chmod -R 777 config storage web/cpresources
+
+      - name: Ping GraphQL endpoint
+        run: |-
+          curl -s -X POST -H "Content-Type: application/json" --data '{ "query": "{ ping }" }' http://nginx/api
+
+      - name: Prepare Gatsby environment
+        run: cp .github/workflows/.env-gatsby headless-front-end/gatsby/.env
+
+      - name: Install Gatsby dependencies
+        run: node --version && cd headless-front-end/gatsby && npm install && npm install gatsby-cli
+
+      - name: Build Gatsby site
+        run: cd headless-front-end/gatsby && npm run build
```

Comment:
I'm pretty sure you can set the working directory underneath the run command.

`working-directory: ./headless-front-end/gatsby`